nixos/lib/eval-config.nix

```diff
@@ -87,16 +83,23 @@ let
     in
       locatedModules ++ legacyModules;
 
-  noUserModules = evalModulesMinimal ({
+  noUserModules = evalSystemConfiguration ({
     inherit prefix specialArgs;
     modules = baseModules ++ extraModules ++ [ pkgsModule modulesModule ];
   });
 
   # Extra arguments that are useful for constructing a similar configuration.
-  modulesModule = {
+  modulesModule = let
+    # keep `mkForce` and similar available to users,
+    # but override what is set in `evalSystemConfiguration`
+    higherThanDefaultPriority = lib.modules.defaultOverridePriority - 1;
+  in {
     config = {
       _module.args = {
-        inherit noUserModules baseModules extraModules modules;
+        inherit extraModules;
+        noUserModules = lib.mkOverride higherThanDefaultPriority noUserModules;
+        baseModules = lib.mkOverride higherThanDefaultPriority baseModules;
+        modules = lib.mkOverride higherThanDefaultPriority modules;
```

Comment:
This is a little disappointing. I was hoping that this file would not have to reimplement these things. It is quite complicated.

Suppose we add `baseModules` to the new function. Wouldn't that allow us to avoid redoing the whole module argument business?
Of course adding that is not great for simplifying the interface, but the distinction between it and `baseModules` is useful for code that needs to interact with specialisations, and some users will need it for that.

Reply:
Avoid this all together only works if we _also_ expose `noUserModules`.

Both would be a _concession_ to path-dependency of the legacy entrypoint and a missed clean-up opportunity.

Since this here would become legacy, that implies, on the long term, it may eventually fall into disuse and go away, the problem of a contrived implementation here also resolves with (enough) patience, alone.

My preference would be to follow through in a green field redesign and _really_ make it unambiguous that `lib` is going to be the _only_ exception to the rule that the module system should be used to configure itself.

Reply:
Perhaps a middle ground is

```
new entrypoint -> simple invocation -> call-nixos-impl.nix
old entrypoint -> more complete invocation -> call-nixos-impl.nix
```

... where we just have to make clear that `call-nixos-impl.nix` is a private interface for the purpose of not duplicating code.

This way we still have the option of removing the old entrypoint and the legacy bits, without incurring the significant cost of having two implementations of the same glue code.

Duplicated glue code is a very real problem. It may not seem bad for glue code to be duplicated as opposed to highly domain-logic-heavy code, but the outcomes are equally bad: stuff breaks when changes are made. Inconsistencies may trickle into the domain where they cause new problems that are then perpetuated because those can't be solved.

That was grim.

Ehm yeah, so I think a shared implementation is really important for the foreseeable future where we do need to maintain both the new and old interfaces.

Reply:
Instead of an actual extra file, we could also hide the legacy support code behind a feature flag. That way we can even enforce that the feature flag is the only way to reach that code. No accidental usages!